### DISCUSSION TITLE
User guide (in PDF) accessible in Trame web application

### DISCUSSION THREAD

**Discussion by karelvrabeckv** [Kitware/trame#472] [Q&A] [answered]
Hello,
I have written a user guide (in PDF) for my web application (developed with the help of Trame). Now, I would like **to add this PDF file to my application** so that every user can **click some button or link**, and **this PDF opens in a new tab** of the web browser (or **it downloads the file**). Actually, I am interested in both ways. **How can I do this in Trame?** I saw some examples with plain text and images, but I was not able to reproduce them.

I am loading the user guide with the help of `LocalFileManager`, but then I am struggling with it. **How can I make the PDF accessible to the users of my application?**

```
import os

from trame.assets.local import LocalFileManager

asset_manager = LocalFileManager(os.path.dirname(__file__))
...
asset_manager.url("user_guide", "./user_guide.pdf")
...
```

Thank you very much! :)

**Comment by jourdain** [comment, marked as answer]
While they are better option, using the LocalFileManager you could do the following

```python
html.A("User Manual", href= asset_manager.user_guide, download=True)
```

A better option would be to create a trame module to serve your pdf over http so you could do

```python
server.enable_module(dict(serve=dict("docs", str(Path(__file__).parent.resolve())))

html.A("User Manual", href="docs/user_guide.pdf", download=True)
```

**Reply by karelvrabeckv** [reply]
Does the better option of enabling module actually work in Docker? I have no problem running locally. Then, the user guide is properly served. But when running in Docker, I got a 404. URL paths are OK. I don't know what I am doing wrong. Is there something explicit to set while running in Docker? Thanks. @jourdain

`self.__server.enable_module({ "serve": { "docs": str(Path(__file__).parent.resolve() / "docs") } })`

**Reply by jourdain** [reply]
In docker, you need to make sure the pdf gets added to the www directory. You can edit the setup directory with some add-on.

**Reply by lahwaacz** [reply]
@jourdain Hmm, relying on Dockerfile to add static files to their right place for deployment sounds like a hack. This should be part of building the application by the provided scripts.

To be more specific, we have the following Dockerfile:
```Dockerfile
FROM kitware/trame:py3.9

COPY --chown=trame-user:trame-user . /local-app
COPY --chown=trame-user:trame-user ./docker /deploy

ENV TRAME_USE_HOST=wss://domain.tld/appname
ENV TRAME_CLIENT_TYPE=vue3

RUN /opt/trame/entrypoint.sh build

COPY --chown=trame-user:trame-user ./appname/app/docs/*.pdf /deploy/server/www/docs/
```

The biggest problem is that the last `COPY` must be placed __after__ the `RUN` command, otherwise `entrypoint.sh` notices that `/deploy/server/www/` already exists/is not empty and does not copy the other static files to the right location. I don't understand why custom static files specified for `"serve"` with `enable_module` are not copied automatically.

I have also other conceptual issues with the Docker deployment (e.g. why does trame _require_ the `/deploy` path in the container to be a mounted volume?) but I don't want to derail the discussion too much...

**Reply by jourdain** [reply]
I didn't tell you to edit the docker file, just to add a directory with your pdf inside `.../setup/www/...` that way the docker image will build and generate its `/deploy/server/www` with your add-on in it.

**Reply by lahwaacz** [reply]
@jourdain The pdfs are not specific to Docker deployment, so they can't be _moved_ into the `.../setup/www/...` directory in the git repo - it would mean that they can't be served when running the application without Docker. Hence, there must be some copy step during the build to _copy_ the pdfs from their _in-source_ location to the _deployment_ location. Please be more explicit in how you recommend to solve this.

**Reply by jourdain** [reply]
Then just use __www_modules__, as described in the [docker doc](https://github.com/Kitware/trame/tree/master/docker#appsyml-mandatory), which is illustrated in the cookie cutter.

**Comment by jourdain** [comment]
Keep in mind that you can list your app [here](https://kitware.github.io/trame/examples/apps/community.html) if that makes sense.
